### PR TITLE
Use test-unit instead of minitest

### PR DESCRIPTION
**Pull request by hsbt**
# Description:

In the Ruby core repository uses test-unit for the test suite. But only rubygems still uses minitest. I hope to switch test-unit from minitest on the rubygems repo.

I need to migrate the following methods:

 * [x] assert_output
 * [x] refute_local_match
 * [x] capture_subprocess_io
 * [x] refute_match with instance object
 * [x] `MiniTest::Mock`

______________

# Tasks:

- [x] Describe the problem / feature
- [ ] Write tests
- [ ] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by djberg96** [comment]
Just curious, what are your thoughts on rspec?

**Comment by hsbt** [comment]
> what are your thoughts on rspec?

The ruby corepository also couldn't use rspec.

**Comment by voxik** [comment]
I thought that the Ruby test-unit is wrapper above minitest and briefly checking the Ruby sources, it still seems to be the case. So is there any plan to change it?

**Comment by deivid-rodriguez** [comment]
Indeed, I didn't actually check this, but now that I check I wonder the same thing.

**Comment by simi** [comment]
I think since ruby still uses test-unit (even it is currently just layer on top of minitest), it would be easier to keep all tests using minitest. Anyway @hsbt would opposite approach work in here as well -> contribute change of test-unit to minitest in ruby repository? That way we don't need to move backwards on rubygems side, but we can actually move forward on ruby side.

**Comment by deivid-rodriguez** [comment]
@simi I think ruby-core is moving the opposite way, since they've been migrating most default gems from minitest to test-unit and their end goal would be to completely get rid of minitest. Please @hsbt correct me if I'm wrong.

In my opinion it's not either going forward or backwards, it's just a different library with very similar syntax. I do think it would make sense that ruby-core and us used `test-unit` since the maintainer of that library is involved in both bundler & ruby-core.

File: lib/rubygems/test_case.rb [hunk 1]
```diff
@@ -140,11 +140,48 @@ def assert_activate(expected, *specs)
...
 
   def assert_directory_exists(path, msg = nil)
-    msg = message(msg) { "Expected path '#{path}' to be a directory" }
-    assert_path_exists path
+    msg = build_message(msg, "Expected path '#{path}' to be a directory")
+    assert_path_exist path
     assert File.directory?(path), msg
   end
 
+  # https://github.com/seattlerb/minitest/blob/master/lib/minitest/assertions.rb#L188
```
[the remark this answers is not in the document]
Comment (hsbt) [reply lib/rubygems/test_case.rb:148]: Good catch, I will update it with the commit hash.

**Comment by hsbt** [comment]
@deivid-rodriguez Thanks for your review! I applied the fixes for your comments.

**Comment by deivid-rodriguez** [comment]
Thanks @hsbt!

**Comment by deivid-rodriguez** [comment]
I have my usual nitpick about changelog :sweat_smile:.

Do you think this qualifies as an "rubygems: enhacement"? In other words, does this affect end users of rubygems at all?

It's a bit unclear whether rubygems base test class is a user facing class or not, given it's shipped inside `lib` as part of rubygems. In my opinion it's not, it's just for internal rubygems testing, so in that case this should not be mentioned in the changelog and be tagged as `rubygems: backport`.

If we consider this a user facing change, then this would technically be a breaking change, because users of this class will now need to have `test-unit` installed rather than `minitest` in order for it to work.

Thoughts?

**Comment by hsbt** [comment]
> In other words, does this affect end users of rubygems at all?

Nothing 😇 

I meant it as "The enhancement for developers" or "Internal changes".

**Comment by deivid-rodriguez** [comment]
Great, then I'll tag it as backport since we usually limit the changelog to end-user facing changes :+1:.

**Comment by deivid-rodriguez** [comment]
So has ruby-core completely moved away from minitest now? :heart_eyes:

**Comment by simi** [comment]
@deivid-rodriguez from my experience, I have never sees `Gem::TestCase` used outside of RubyGems itself. Is there any particular reason to keep it in `lib` actually? I think that pattern is useful when you would like to expose it to users, but in this case it is really just internal test helper.

**Comment by deivid-rodriguez** [comment]
Yeah, me neither. I don't know why it's exposed in `lib`, I think we should move it to `test` to make it clear that it's not supposed to be exposed to end users. And moving it to `test` would also make things easier for OS packagers, since we would eliminate the need for cross-referencing `lib` from `test`, which is particularly problematic for them.

**Comment by deivid-rodriguez** [comment]
I'm dusting off an old branch where I tried to do just that, and I'm noticing that it indeed seems like `Gem::TestCase` is indeed expected to be exposed to end users, since there are some comments in some methods there saying things like

>  This method is available when requiring 'rubygems/test_case'

Maybe for testing rubygems plugins or something like that? :man_shrugging:

Anyways, the approach my branch was taking was to copy over the files to `test` and add a deprecation on top to the ones inside `lib`. But I guess since we're doing this, we can go without the deprecation and find out whether this breaks anything for someone?

**Comment by deivid-rodriguez** [comment]
I opened https://github.com/rubygems/rubygems/pull/4599 just to make the approach I'm talking about more clear. I tend to think it's fine to just move forward and remove the files since none of us have actually see any usage of the `Gem::TestCase` in the realworld :man_shrugging:.

**Pull request merged**